### PR TITLE
[bug fix] Correction of fresnel_polarized for complex IORs (metallic)

### PR DESCRIPTION
**Pull request by shinyoung-yi**
Fix the imaginary part of cosine refracted angles for complex IORs (metallic) in `fresnel_polarized()`

### 1. Motivation
First of all, note that there are two conventions: $\exp\left(+i\omega t\right)$ and $\exp\left(-i\omega t\right)$ formulations to express waves. Mitsuba 3 uses the former one ( $\exp\left(+i\omega t\right)$ ), which yields $\eta = n - ik$ formulation for complex refractive indices.
I would not discuss this background since it has already been discussed in Issue #1113 and PR #1150 with @tizian .

This PR fixes the remained problem raised in Issue #1113 . The current Mitsuba 3 produces wrong sign of imaginary parts of $\cos\tilde{\theta}_t$ ($\tilde\theta_t$ denotes a complex refracted angle) for complex IORs.

Concretely, with the $\exp\left(+i\omega t\right)$ convention, my claim is that $\Im \cos \tilde\theta_t \le 0$, as best of my knowledge.
However current implementation of `fresnel_polarized(Float cos_theta_i, dr::Complex<Float> eta)` produces inconsistent signs for `cos_theta_t` depending on ranges of arguments.

While a technical reasoning for my claim $\Im \cos \tilde\theta_t \le 0$ is contained in Issue #1113, I would like to focus on concrete outputs for `fresnel_polarized()` here.

### 2. Discontinuous outputs for `fresnel_polarized`
```python
import numpy as np
import matplotlib.pyplot as plt
import drjit as dr
import mitsuba as mi
mi.set_variant('cuda_ad_rgb')

print(f"{dr.__version__ = }")
print(f"{mi.__version__ = }")

print(mi.fresnel_polarized(0.5, 0.5))
print(mi.fresnel_polarized(0.5, 0.5-1e-8j))
```
As the above code, let's imagine Fresnel refectance values for slightly different two refractive indices.
It will be nature that two results would be almost equal, but the current Mitsuba 3 implementation produces sign flippling for `a_s` and `a_p` as follows.
**BEFORE:**
```
([[-0.3333333432674408 + 0.9428090453147888i]], [[-0.9393939971923828 + 0.34283965826034546i]], [0.0], [[0.5 + 0.0i]], [[2.0 - 0.0i]])
([[-0.3333333432674408 - 0.9428090453147888i]], [[-0.9393939971923828 - 0.34283965826034546i]], [0.0], [[0.5 - 9.99999993922529e-09i]], [[2.0 + 3.999999975690116e-08i]])
```
After bug fix by this PR, the results have changed as follows.
**AFTER:**
```
([[-0.3333333432674408 + 0.9428090453147888i]], [[-0.9393939971923828 + 0.34283965826034546i]], [0.0], [[0.5 + 0.0i]], [[2.0 - 0.0i]])
([[-0.3333333432674408 + 0.9428090453147888i]], [[-0.9393939971923828 + 0.34283962845802307i]], [0.0], [[0.5 - 9.99999993922529e-09i]], [[2.0 + 3.999999975690116e-08i]])
```

### 3. Reproducing plots in [the documentation](https://mitsuba.readthedocs.io/en/latest/src/key_topics/polarization.html#analysis-and-mueller-matrix-formulation)
I also provide a verification through reproducing plots of reflectance and phase delay in the Mitsuba 3 documentation.
Mitsuba function `fresnel_polarized()` does not returns complex values of `cos_theta_t` (it returns only the real part), so also implemented a short numpy code to show $\Im \cos\theta_t\le 0$ is correct and $\Im \cos\theta_t\ge 0$ is not relavant to the convention which Mitsuba 3 relies on.

```python
N_theta = 91

theta_i = dr.linspace(mi.Float, 0, 90, N_theta)
cos_theta_i = dr.cos(dr.deg2rad(theta_i))
sinsq_i = 1 - cos_theta_i.numpy()**2

def fresnel_reimpl(cos_i, cos_t, eta):
    rs = (cos_i - eta*cos_t) / (cos_i + eta*cos_t)
    rp = (eta*cos_i - cos_t) / (eta*cos_i + cos_t)
    return rs, rp

def show_fresnel(eta, suptitle=None):
    cos_t = np.sqrt(1+0j - sinsq_i / (eta**2))
    cos_t_ipos = np.where(cos_t.imag < 0, -cos_t, cos_t)
    cos_t_ineg = np.where(cos_t.imag > 0, -cos_t, cos_t)
    assert (cos_t_ipos.imag >= 0).all()
    assert (cos_t_ineg.imag <= 0).all()

    configs = [r"$\Im \cos \theta_t > 0$", r"$\Im \cos \theta_t < 0$", "Mitsuba 3"]
    fig, axes = plt.subplots(3, len(configs), figsize=(9, 8))
    for i, label in enumerate(configs): # for each column
        if i == 0:
            rs, rp = fresnel_reimpl(cos_theta_i.numpy(), cos_t_ipos, eta)
        elif i == 1:
            rs, rp = fresnel_reimpl(cos_theta_i.numpy(), cos_t_ineg, eta)
        else:
            rs, rp, _, _, _ = mi.fresnel_polarized(cos_theta_i, eta)
            rs = rs.numpy()
            # print(f"{rs.shape = }")
            if dr.is_complex_v(rs):
                rs = rs[:,0] + rs[:,1] * 1j
            rp = rp.numpy()
            if dr.is_complex_v(rp):
                rp = rp[:,0] + rp[:,1] * 1j

        # ==================== First row ====================
        axes[0, i].plot(theta_i, rs.real, label=r"$\Re r_s$")
        axes[0, i].plot(theta_i, rs.imag, label=r"$\Im r_s$")
        axes[0, i].plot(theta_i, rp.real, label=r"$\Re r_p$")
        axes[0, i].plot(theta_i, rp.imag, label=r"$\Im r_p$")
        axes[0, i].set_xlabel(r"$\theta_i$")
        axes[0, i].set_ylabel("reflectance (complex amplitude)")
        axes[0, i].set_title(label)
        axes[0, i].legend()

        # ==================== Second row ====================
        Rs = np.abs(rs*rs)
        Rp = np.abs(rp*rp)

        axes[1, i].plot(theta_i, Rs, label=r"$R_s$")
        axes[1, i].plot(theta_i, Rp, label=r"$R_p$")
        axes[1, i].set_xlabel(r"$\theta_i$")
        axes[1, i].set_ylabel("reflectance (power)")
        axes[1, i].set_title(label)
        axes[1, i].legend()

        # ==================== Third row ====================
        delta = np.angle(rp/rs, deg=True)
        # print(f"{delta.shape = }")
        if delta[0] < -150 and delta[1] > 150:
            # For better visualization
            delta[0] += 360
        axes[2, i].plot(theta_i, delta)
        axes[2, i].set_xlabel(r"$\theta_i$")
        axes[2, i].set_ylabel(r"$\delta_p - \delta_s$")
        axes[2, i].set_title(label)
    if not suptitle is None:
        plt.suptitle(suptitle)
    plt.tight_layout()
```

```python
eta = 1/1.5
show_fresnel(eta, "$\\eta=1/1.5$")
```
Before eta=0 67
**Figure 1.**

Let's compare it with *Figure 12* in [the documentation](https://mitsuba.readthedocs.io/en/latest/src/key_topics/polarization.html). From the first and second column in **Figure 1**, we can see $\Im \cos\theta_t \le 0$ produces the same plots as the documentation. Current Mitsuba 3 implementaion of `fresnel_polarized(Float, Float)` also produces the same results as the second column, so I did not modified `fresnel_polarized(Float, Float)`. **Figure 1** does not changed by this PR.

On the other hand, for another overloaded function `fresnel_polarized(Float, dr::Complex<Float>)`, it yields wrong values but this error is somewhat obscure since it produces correct value for `0.183-3.43j` but incorrect one for `0.5-1e-8j`.
I provide a diagram to show how the current implementation has a bug.
complex fresnel range
**Figure 2.**

(a) shows ranges of the complex IOR. By our convention, we focus on the region of orange and red colors.
(I consider we need a much more sophisticated careful discussion for signs for complex IORs and $\cos\theta_t$ with a negative real part of a IOR (Sec. 17.3.1 in [Modern Electrodynamics - Andrew Zangwill]). However, I cannot handle such cases due to my limited knowledge. Thus I would like to focus on cases with $\Re\eta \ge 0$, which covers all materials implemented in Mitsuba 3)
In (a) there are two markers for specific IORs: i) and ii). I will show plots like **Figure 1** for cases of the two markers later.

(b) shows $\cos^2\theta_t$. Note that due to the range of $\sin\theta_i$ (0 to 1), specific IOR values i) and ii) have been changed into lines. For i) $\eta=0.5-10^{-8}i$, a circle marker indicates a specific value of $\cos\theta_i=0.5$, which is the case reported above (2. Discontinuous outputs for `fresnel_polarized`).

(c) indicates correct values of $\cos\theta_t$. Choosing the $\exp\left(+i\omega t\right)$ convention, we should choose the lower half-infinite red line on $\Re = 0$. 

While directly applying `sqrt` (in Dr.Jit and most of other scientific computing library), the resulting value shown in (d) has wrong sign of the imaginary part when $\Re = 0$. It can be corrected by (f) (= this PR).

However, current Mitsuba 3 implementation uses (e), which produces incorrect values depending on arguments. For instance, the IOR ii) $\eta=0.183-3.43i$ (*Figure 14* of the documentation) yields correct values.
```python
eta = 0.183 - 3.43j
show_fresnel(eta, "$\\eta=0.183-3.43i$")
# plt.savefig(f"Before eta={eta:.2f}.png")
plt.savefig(f"After eta={eta:.2f}.png")
```
Before eta=0 18-3 43j
**Figure 3.** (Applying this PR still yields the same result)

However, the IOR ii) $\eta=0.5-10^{-8}i$ yields flipped signs in $\Im \cos\theta_t$.
```python
eta = 0.5-1e-8j
show_fresnel(eta, "$\\eta=0.5-10^{-8}i$")
```
Compare eta=0 50-0 00j
**Figure 4.** (Applying this commit changes the results in the third column. I manually attached the third column of the changed result as the forth column of this figure.)

Though this PR, `fresnel_polarized` will produces correct values of $\cos\theta_t$ with consistent negative signs for imaginary parts.

## Checklist

- [x] My code follows the [style guidelines](https://mitsuba.readthedocs.io/en/latest/src/developer_guide.html#coding-style) of this project
- [x] My changes generate no new warnings
- [x] My code also compiles for `cuda_*` and `llvm_*` variants. If you can't test this, please leave below
- [ ] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I cleaned the commit history and removed any "Merge" commits
- [x] I give permission that the Mitsuba 3 project may redistribute my contributions under the terms of its [license](https://github.com/mitsuba-renderer/mitsuba/blob/master/LICENSE)

### CODE REVIEW

**Comment by tizian** [comment]
Hi Shinyoung,
Sorry for the very long delay on this but I finally took a careful look at this issue. I also ran some local experiments and test renderings and agree with the discussion above. Thanks for the super thorough investigation!

**Comment by wjakob** [comment]
One question to @shinyoung-yi and @tizian : with this change applied, two polarization-related testcases fail.

``test05_specular_reflection`` and ``test06_phase_tir``. Does the investigation here imply that these two testcases are incorrect? If so, it would be important to also update the tests with this new informatoin.

**Comment by shinyoung-yi** [comment]
I've briefly took a look for the testing scripts. @wjakob, you're referring to`src/render/tests/test_mueller.py:test05_specular_reflection` and `src/render/tests/test_fresnel.py:test06_phase_tir`, right?

The failures in these test cases seem not to be what I intended, as I was focusing on fixing cases of complex IoR (metallic), while `test05_specular_reflection` and `test06_phase_tir` appear to be testing only real IoRs.

I need to revisit the details and run the testing scripts myself.
This may take a couple of days.

**Comment by tizian** [comment]
Ok, I understand what was the confusion now. @shinyoung-yi and I both ran test plots in the LLVM/CUDA backends and those had an unfortunate inconsistency in how they compute the complex root of a real number, depending on the +/- sign of the `0j` component.
This is why the plots looked correct but the unit tests (running the scalar backend) started to fail.

Example:
```
mi.set_variant('llvm_ad_rgb')
z = mi.Complex2f(-0.5, +0.0)
print(dr.sqrt(z).imag)
z = mi.Complex2f(-0.5, -0.0)
print(dr.sqrt(z).imag)

# output:
# [0.707107]
# [0.707107]

mi.set_variant('scalar_rgb')
z = mi.Complex2f(-0.5, +0.0)
print(dr.sqrt(z).imag)
z = mi.Complex2f(-0.5, -0.0)
print(dr.sqrt(z).imag)

# output:
# 0.7071067690849304
# -0.7071067690849304
```
@wjakob already pushed a fix to this to DrJIT (https://github.com/mitsuba-renderer/drjit/pull/300) 🙇 

With this patch in place, the suggested fix in this PR is still a bit problematic and leads to wrong results for real IORs.
For clarity I'll go through both options again, and finally propose a working solution. Here is a slightly adjusted plotting script from above for reproducibility:
```python
import numpy as np
import matplotlib.pyplot as plt
import mitsuba as mi
import drjit as dr
mi.set_variant('llvm_ad_rgb')

N_theta = 91
theta_i = dr.linspace(mi.Float, 1e-3, 90-1e-3, N_theta)
cos_theta_i = dr.cos(dr.deg2rad(theta_i))
sinsq_i = 1 - cos_theta_i.numpy()**2

def fresnel_reimpl(cos_i, cos_t, eta):
    rs = (cos_i - eta*cos_t) / (cos_i + eta*cos_t)
    rp = (eta*cos_i - cos_t) / (eta*cos_i + cos_t)
    return rs, rp

def show_fresnel(eta, suptitle=None):
    cos_t = np.sqrt(1+0j - sinsq_i / (eta**2))
    cos_t_ipos = np.where(cos_t.imag < 0, -cos_t, cos_t)
    cos_t_ineg = np.where(cos_t.imag > 0, -cos_t, cos_t)
    assert (cos_t_ipos.imag >= 0).all()
    assert (cos_t_ineg.imag <= 0).all()

    configs = [r"$\Im \cos \theta_t > 0$", r"$\Im \cos \theta_t < 0$", "Mitsuba 3"]
    _, axes = plt.subplots(1, len(configs), figsize=(9, 3))
    for i, label in enumerate(configs): # for each column
        if i == 0:
            rs, rp = fresnel_reimpl(cos_theta_i.numpy(), cos_t_ipos, eta)
        elif i == 1:
            rs, rp = fresnel_reimpl(cos_theta_i.numpy(), cos_t_ineg, eta)
        else:
            rs, rp, _, _, _ = mi.fresnel_polarized(cos_theta_i, mi.Complex2f(eta))
            rs = rs.numpy()
            if dr.is_complex_v(rs):
                rs = rs[:,0] + rs[:,1] * 1j
            rp = rp.numpy()
            if dr.is_complex_v(rp):
                rp = rp[:,0] + rp[:,1] * 1j

        delta = np.angle(rp/rs, deg=True)
        delta[delta < 0] += 360 # For better visualization
        axes[i].plot(theta_i.numpy(), delta)
        axes[i].set_xlabel(r"$\theta_i$")
        axes[i].set_ylabel(r"$\delta_p - \delta_s$")
        axes[i].set_title(label)
        axes[i].set_xlim([0, 90])
        axes[i].set_xticks([0, 45, 90])
        axes[i].set_ylim([-10, 370])
        axes[i].set_yticks([0, 90, 180, 270, 360])
    if not suptitle is None:
        plt.suptitle(suptitle)
    plt.tight_layout()
```

I ran the same representative examples from above:
```python
eta = 0.183 - 3.43j
show_fresnel(eta, "$\\eta=0.183-3.43i$")

eta = 0.5-1e-8j
show_fresnel(eta, "$\\eta=0.5-10^{-8}i$")

eta = 1/1.5
show_fresnel(eta, "$\\eta=1/1.5$")
```

## Option 1 (original code)
```cpp
dr::Complex<Float> cos_theta_t = dr::sqrt(cos_theta_t_sqr);
cos_theta_t = dr::mulsign(dr::Array<Float, 2>(cos_theta_t), dr::real(cos_theta_t_sqr));
```

This is incorrect for the dielectric + tiny imaginary epsilon test case (third row).

v1_conductor
v1_dielectric
v1_epsconductor


## Option 2 (this PR)
```cpp
dr::Complex<Float> cos_theta_t = dr::conj(dr::sqrt(dr::conj(cos_theta_t_sqr)));
```

This is incorrect in the purely dielectric case (second row).
v2_conductor
v2_dielectric
v2_epsconductor

## Option 3 (new suggestion)
I think the most straightforward option is to simply be very explicit that we want to pick the complex root that has a _negative imaginary component_:
```cpp
dr::Complex<Float> cos_theta_t = dr::sqrt(cos_theta_t_sqr);
dr::masked(cos_theta_t, dr::imag(cos_theta_t) > 0) = dr::conj(cos_theta_t);
```

This gives the correct phase shift in all three cases.
v3_conductor
v3_dielectric
v3_epsconductor

With this, the mentioned to tests in `test_fresnel.py` and `test_mueller.py` also all pass.

Here is the final diff I'm suggesting that consistently does this also in the real-IOR implementation (though the corner cases above can technically never happen there):

```diff
diff --git a/include/mitsuba/render/fresnel.h b/include/mitsuba/render/fresnel.h
index e706a8f1..48bbb56b 100644
--- a/include/mitsuba/render/fresnel.h
+++ b/include/mitsuba/render/fresnel.h
@@ -166,7 +166,8 @@ fresnel_polarized(Float cos_theta_i, Float eta) {
     /* Choose the appropriate sign of the root (important when computing the
        phase difference under total internal reflection, see appendix A.2 of
        "Stellar Polarimetry" by David Clarke) */
-    cos_theta_t = dr::mulsign(dr::Array<Float, 2>(cos_theta_t), cos_theta_t_sqr);
+    // cos_theta_t = dr::mulsign(dr::Array<Float, 2>(cos_theta_t), cos_theta_t_sqr);
+    dr::masked(cos_theta_t, dr::imag(cos_theta_t) > 0) = dr::conj(cos_theta_t);

     /* Amplitudes of reflected waves. The sign of 'a_p' used here is referred
        to as the "Verdet convention" which more common in the literature
@@ -248,7 +249,7 @@ fresnel_polarized(Float cos_theta_i, dr::Complex<Float> eta) {
     /* Choose the appropriate sign of the root (important when computing the
        phase difference under total internal reflection, see appendix A.2 of
        "Stellar Polarimetry" by David Clarke) */
-    cos_theta_t = dr::mulsign(dr::Array<Float, 2>(cos_theta_t), dr::real(cos_theta_t_sqr));
+    dr::masked(cos_theta_t, dr::imag(cos_theta_t) > 0) = dr::conj(cos_theta_t);

     /* Amplitudes of reflected waves. The sign of 'a_p' used here is referred
        to as the "Verdet convention" which more common in the literature
```

**Comment by tizian** [comment]
@shinyoung-yi Could you please update the PR with those changes? It might be possible for me to push directly but I don't want to force-push to your own fork's master branch.

**Comment by shinyoung-yi** [comment]
Thanks to @tizian for the analysis. I totally agree with you.
I'll be able to update the PR in about 7-10 hours, once my personal schedule is wrapped up.

**Comment by shinyoung-yi** [comment]
Sorry for late. Now I update this PR based on @tizian 's correction.
I tested that it reproduces all the (correct version of) figures discussed above and passes the two test scripts!
```
(mitsuba_fork) D:\AcademicSoftwares\differentiable_renders\mitsuba3_fork>pytest src/render/tests
/test_mueller.py
Running the full test suite. To skip slow tests, please run 'pytest -m "not slow"'
===================================== test session starts =====================================
platform win32 -- Python 3.11.10, pytest-8.3.3, pluggy-1.5.0
rootdir: D:\AcademicSoftwares\differentiable_renders\mitsuba3_fork
configfile: pyproject.toml
collected 9 items

src\render\tests\test_mueller.py .........                                               [100%]

====================================== 9 passed in 0.15s ======================================

(mitsuba_fork) D:\AcademicSoftwares\differentiable_renders\mitsuba3_fork>pytest src/render/tests/test_fresnel.py
Running the full test suite. To skip slow tests, please run 'pytest -m "not slow"'
===================================== test session starts =====================================
platform win32 -- Python 3.11.10, pytest-8.3.3, pluggy-1.5.0
rootdir: D:\AcademicSoftwares\differentiable_renders\mitsuba3_fork
configfile: pyproject.toml
collected 11 items

src\render\tests\test_fresnel.py ...........                                             [100%]

===================================== 11 passed in 0.56s ======================================
```

**Comment by wjakob** [comment]
Thank you very much 🙇
